File: ios/Shared/Sources/Features/LmsAssignmentReference.swift
import Foundation
import MoodleNativeCore
import MoodleNativeNetworking

public struct LmsAssignmentReference: Hashable, Identifiable, Sendable {
    public let assignmentID: Int
    public let courseModuleID: Int
    public let courseID: Int?
    public let courseTitle: String?
    public let courseShortName: String?
    public let title: String
    public let introPreview: String?
    public let dueDate: Date?
    public let allowsSubmissionsFromDate: Date?
    public let cutoffDate: Date?
    public let updatedAt: Date?
    public let submissionDrafts: Bool?
    public let requiresSubmissionStatement: Bool?
    public let submissionStatement: String?
    public let timeLimit: Int?
    public let enabledSubmissionPluginTypes: [String]?
    public let onlineTextWordLimit: Int?
    public let detailURL: URL?
    public let completionRequirements: [LmsModuleCompletionRequirement]

    public init(
        assignmentID: Int,
        courseModuleID: Int,
        courseID: Int?,
        courseTitle: String?,
        courseShortName: String?,
        title: String,
        introPreview: String?,
        dueDate: Date?,
        allowsSubmissionsFromDate: Date?,
        cutoffDate: Date?,
        updatedAt: Date?,
        submissionDrafts: Bool? = nil,
        requiresSubmissionStatement: Bool? = nil,
        submissionStatement: String? = nil,
        timeLimit: Int? = nil,
        enabledSubmissionPluginTypes: [String]? = nil,
        onlineTextWordLimit: Int? = nil,
        detailURL: URL?,
        completionRequirements: [LmsModuleCompletionRequirement]
    ) {
        self.assignmentID = assignmentID
        self.courseModuleID = courseModuleID
        self.courseID = courseID
        self.courseTitle = courseTitle
        self.courseShortName = courseShortName
        self.title = title
        self.introPreview = introPreview
        self.dueDate = dueDate
        self.allowsSubmissionsFromDate = allowsSubmissionsFromDate
        self.cutoffDate = cutoffDate
        self.updatedAt = updatedAt
        self.submissionDrafts = submissionDrafts
        self.requiresSubmissionStatement = requiresSubmissionStatement
        self.submissionStatement = submissionStatement
        self.timeLimit = timeLimit
        self.enabledSubmissionPluginTypes = enabledSubmissionPluginTypes
        self.onlineTextWordLimit = onlineTextWordLimit
        self.detailURL = detailURL
        self.completionRequirements = completionRequirements
    }

    public var id: Int {
        assignmentID
    }

    public var displayCourseTitle: String? {
        guard let courseTitle else {
            return nil
        }

        let trimmed = courseTitle.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    public var hasOverviewDetails: Bool {
        if allowsSubmissionsFromDate != nil || dueDate != nil || cutoffDate != nil {
            return true
        }

        guard let introPreview else {
            return false
        }

        return introPreview.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
    }

    public func merging(courseModule: LmsWebServiceClient.CourseModule) -> LmsAssignmentReference {
        let mergedDetailURL = detailURL ?? courseModule.url.flatMap(URL.init(string:))
        let mergedRequirements =
            completionRequirements.isEmpty
            ? courseModule.completionData?.requirements ?? []
            : completionRequirements

        return LmsAssignmentReference(
            assignmentID: assignmentID,
            courseModuleID: courseModuleID,
            courseID: courseID,
            courseTitle: courseTitle,
            courseShortName: courseShortName,
            title: title,
            introPreview: introPreview,
            dueDate: dueDate,
            allowsSubmissionsFromDate: allowsSubmissionsFromDate,
            cutoffDate: cutoffDate,
            updatedAt: updatedAt,
            submissionDrafts: submissionDrafts,
            requiresSubmissionStatement: requiresSubmissionStatement,
            submissionStatement: submissionStatement,
            timeLimit: timeLimit,
            enabledSubmissionPluginTypes: enabledSubmissionPluginTypes,
            onlineTextWordLimit: onlineTextWordLimit,
            detailURL: mergedDetailURL,
            completionRequirements: mergedRequirements
        )
    }
}

public extension LmsAssignmentItem {
    func reference(siteURL: String) -> LmsAssignmentReference {
        LmsAssignmentReference(
            assignmentID: id,
            courseModuleID: courseModuleID,
            courseID: courseID,
            courseTitle: courseTitle,
            courseShortName: courseShortName,
            title: title,
            introPreview: introPreview,
            dueDate: dueDate,
            allowsSubmissionsFromDate: allowsSubmissionsFromDate,
            cutoffDate: cutoffDate,
            updatedAt: updatedAt,
            submissionDrafts: submissionDrafts,
            requiresSubmissionStatement: requiresSubmissionStatement,
            submissionStatement: submissionStatement,
            timeLimit: timeLimit,
            enabledSubmissionPluginTypes: enabledSubmissionPluginTypes,
            onlineTextWordLimit: onlineTextWordLimit,
            detailURL: URL.assignmentDetail(baseSiteURL: siteURL, courseModuleID: courseModuleID),
            completionRequirements: []
        )
    }
}

extension URL {
    fileprivate static func assignmentDetail(baseSiteURL: String, courseModuleID: Int) -> URL? {
        guard var components = URLComponents(string: baseSiteURL) else {
            return nil
        }

        components.path = "/mod/assign/view.php"
        components.queryItems = [URLQueryItem(name: "id", value: String(courseModuleID))]
        return components.url
    }
}

File: android/features/src/main/kotlin/dev/example/moodlenative/features/LmsAssignmentReference.kt
package dev.example.moodlenative.features

import dev.example.moodlenative.core.LmsAssignmentItem
import java.net.URI
import java.time.Instant

data class LmsAssignmentReference(
    val assignmentID: Int,
    val courseModuleID: Int,
    val courseID: Int?,
    val courseTitle: String?,
    val courseShortName: String?,
    val title: String,
    val introPreview: String?,
    val dueDate: Instant?,
    val allowsSubmissionsFromDate: Instant?,
    val cutoffDate: Instant?,
    val updatedAt: Instant?,
    val submissionDrafts: Boolean? = null,
    val requiresSubmissionStatement: Boolean? = null,
    val submissionStatement: String? = null,
    val timeLimit: Int? = null,
    val enabledSubmissionPluginTypes: List<String>? = null,
    val onlineTextWordLimit: Int? = null,
    val detailURL: URI?,
    val completionRequirements: List<LmsModuleCompletionRequirement> = emptyList(),
) {
    val id: Int
        get() = assignmentID

    val displayCourseTitle: String?
        get() = courseTitle?.trim()?.takeIf { it.isNotEmpty() }

    val hasOverviewDetails: Boolean
        get() = allowsSubmissionsFromDate != null ||
            dueDate != null ||
            cutoffDate != null ||
            introPreview?.trim()?.isNotEmpty() == true

    fun merging(courseModule: LmsCourseModule): LmsAssignmentReference {
        val mergedDetailURL = detailURL ?: courseModule.url?.let { url ->
            runCatching { URI(url) }.getOrNull()
        }
        val mergedRequirements = if (completionRequirements.isEmpty()) {
            courseModule.completionRequirements
        } else {
            completionRequirements
        }

        return copy(
            detailURL = mergedDetailURL,
            completionRequirements = mergedRequirements,
        )
    }
}

fun LmsAssignmentItem.reference(siteURL: String): LmsAssignmentReference =
    LmsAssignmentReference(
        assignmentID = id,
        courseModuleID = courseModuleID,
        courseID = courseID,
        courseTitle = courseTitle,
        courseShortName = courseShortName,
        title = title,
        introPreview = introPreview,
        dueDate = dueDate,
        allowsSubmissionsFromDate = allowsSubmissionsFromDate,
        cutoffDate = cutoffDate,
        updatedAt = updatedAt,
        submissionDrafts = submissionDrafts,
        requiresSubmissionStatement = requiresSubmissionStatement,
        submissionStatement = submissionStatement,
        timeLimit = timeLimit,
        enabledSubmissionPluginTypes = enabledSubmissionPluginTypes,
        onlineTextWordLimit = onlineTextWordLimit,
        detailURL = assignmentDetailURL(baseSiteURL = siteURL, courseModuleID = courseModuleID),
        completionRequirements = emptyList(),
    )

private fun assignmentDetailURL(baseSiteURL: String, courseModuleID: Int): URI? {
    val baseURL = runCatching { URI(baseSiteURL.trim()) }.getOrNull() ?: return null
    if (baseURL.scheme.isNullOrEmpty() || baseURL.authority.isNullOrEmpty()) {
        return null
    }

    return runCatching {
        URI("${baseURL.scheme}://${baseURL.authority}/mod/assign/view.php?id=$courseModuleID")
    }.getOrNull()
}
